include/materials/norsand.tcc

```diff
@@ -114,54 +116,26 @@ bool mpm::NorSand<Tdim>::compute_elastic_tensor() {
 
 //! Compute stress invariants
 template <unsigned Tdim>
-Eigen::Matrix<double, 6, 1> mpm::NorSand<Tdim>::compute_stress_invariants(
+double* mpm::NorSand<Tdim>::compute_stress_invariants(
     const Vector6d& stress, mpm::dense_map* state_vars) {
 
   // Note that in this subroutine, stress is compression positive
 
+  // Store to return
+  static double invariants[4];
+
   // Compute mean stress p
-  double mean_p = (stress(0) + stress(1) + stress(2)) / 3.;
-  mean_p = check_low(mean_p);
-
-  // Compute J2
-  double j2 = (std::pow((stress(0) - stress(1)), 2) +
-               std::pow((stress(1) - stress(2)), 2) +
-               std::pow((stress(0) - stress(2)), 2)) /
-                  6.0 +
-              std::pow(stress(3), 2) + std::pow(stress(4), 2) +
-              std::pow(stress(5), 2);
-  j2 = check_low(j2);
+  invariants[0] = check_low(-1. * mpm::materials::p(-stress));
```

Comment:
Invariants can be returned in any order, so please use actual pointers to individual invariants

Reply:
I could not imagine this, could you give an example? Like shared pointer?